File: Tools/verify_tree.swift
#!/usr/bin/env swift
//
// verify_tree.swift
// Load CMUs from bin file, build tree, and verify root matches chain
//

import Foundation

// Load the FFI
@_silgen_name("zipherx_tree_init")
func treeInit() -> Bool

@_silgen_name("zipherx_tree_append")
func treeAppend(_ cmu: UnsafePointer<UInt8>) -> UInt64

@_silgen_name("zipherx_tree_root")
func treeRoot(_ rootOut: UnsafeMutablePointer<UInt8>) -> Bool

@_silgen_name("zipherx_tree_size")
func treeSize() -> UInt64

func main() {
    print("🌳 Tree Verification Tool")
    print("========================\n")

    // Load CMUs from file
    let inputPath = FileManager.default.currentDirectoryPath + "/commitment_tree.bin"

    guard let data = FileManager.default.contents(atPath: inputPath) else {
        print("❌ Could not read \(inputPath)")
        return
    }

    // Parse: [count: UInt64][cmu1: 32 bytes][cmu2: 32 bytes]...
    var offset = 0
    let count = data.withUnsafeBytes { ptr -> UInt64 in
        ptr.load(fromByteOffset: 0, as: UInt64.self)
    }
    offset = 8

    print("📦 Loading \(count) CMUs from file...")

    // Initialize tree
    guard treeInit() else {
        print("❌ Failed to init tree")
        return
    }

    // Append all CMUs
    let startTime = Date()
    var position: UInt64 = 0

    for i in 0..<count {
        let cmuData = data.subdata(in: offset..<(offset + 32))
        offset += 32

        cmuData.withUnsafeBytes { ptr in
            position = treeAppend(ptr.baseAddress!.assumingMemoryBound(to: UInt8.self))
        }

        if i % 100000 == 0 {
            let progress = Double(i) / Double(count) * 100
            print("\r⏳ Progress: \(String(format: "%.1f", progress))% (\(i)/\(count))", terminator: "")
            fflush(stdout)
        }
    }

    let elapsed = Date().timeIntervalSince(startTime)
    print("\n✅ Loaded \(count) CMUs in \(String(format: "%.1f", elapsed))s")

    // Get tree root
    var root = [UInt8](repeating: 0, count: 32)
    guard treeRoot(&root) else {
        print("❌ Failed to get tree root")
        return
    }

    // Convert to hex (note: the root needs to be reversed for comparison with RPC)
    let rootHex = root.reversed().map { String(format: "%02x", $0) }.joined()

    print("\n📊 Tree Statistics:")
    print("   Size: \(treeSize()) commitments")
    print("   Root: \(rootHex)")

    print("\n💡 Compare with chain's finalsaplingroot using:")
    print("   zclassic-cli getblockheader $(zclassic-cli getblockhash HEIGHT) true | grep finalsaplingroot")
}

main()

File: Libraries/zipherx-ffi/src/bin/verify_tree.rs
//! Verify commitment tree by computing root from CMU file
//!
//! Usage: cargo run --bin verify_tree /path/to/commitment_tree.bin

use std::env;
use std::fs::File;
use std::io::{Read, BufReader};
use incrementalmerkletree::frontier::CommitmentTree;
use zcash_primitives::sapling::Node;
use zcash_primitives::merkle_tree::HashSer;

fn main() {
    let args: Vec<String> = env::args().collect();
    let path = if args.len() > 1 {
        &args[1]
    } else {
        "commitment_tree.bin"
    };

    println!("🌳 Tree Verification Tool");
    println!("========================\n");
    println!("📁 Loading: {}", path);

    let file = match File::open(path) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("❌ Failed to open file: {}", e);
            return;
        }
    };

    let mut reader = BufReader::new(file);

    // Read count (8 bytes little-endian)
    let mut count_buf = [0u8; 8];
    if reader.read_exact(&mut count_buf).is_err() {
        eprintln!("❌ Failed to read count");
        return;
    }
    let count = u64::from_le_bytes(count_buf);
    println!("📦 CMU count: {}", count);

    // Initialize tree
    let mut tree: CommitmentTree<Node, 32> = CommitmentTree::empty();

    // Read and append all CMUs
    let mut cmu_buf = [0u8; 32];
    let mut processed = 0u64;

    println!("⏳ Building tree...");

    loop {
        match reader.read_exact(&mut cmu_buf) {
            Ok(_) => {
                // Parse CMU as Node
                let node = match Node::read(&cmu_buf[..]) {
                    Ok(n) => n,
                    Err(e) => {
                        eprintln!("❌ Failed to parse CMU at position {}: {:?}", processed, e);
                        return;
                    }
                };

                // Append to tree
                if tree.append(node).is_err() {
                    eprintln!("❌ Failed to append CMU at position {}", processed);
                    return;
                }

                processed += 1;

                if processed % 100000 == 0 {
                    let progress = (processed as f64 / count as f64) * 100.0;
                    eprint!("\r⏳ Progress: {:.1}% ({}/{})", progress, processed, count);
                }
            }
            Err(_) => break,
        }
    }

    eprintln!("\r✅ Loaded {} CMUs                    ", processed);

    if processed != count {
        eprintln!("⚠️  Warning: Expected {} CMUs but read {}", count, processed);
    }

    // Get tree root
    let root = tree.root();

    // Convert to bytes and reverse for display (RPC shows big-endian)
    let mut root_bytes = [0u8; 32];
    root.write(&mut root_bytes[..]).unwrap();
    root_bytes.reverse();

    let root_hex: String = root_bytes.iter().map(|b| format!("{:02x}", b)).collect();

    println!("\n📊 Tree Statistics:");
    println!("   Size: {} commitments", processed);
    println!("   Root: {}", root_hex);

    println!("\n💡 Compare with chain's finalsaplingroot:");
    println!("   zclassic-cli getblockheader $(zclassic-cli getblockhash HEIGHT) true | grep finalsaplingroot");
}
